

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>stlearn.preprocessing.log_scale &mdash; stLearn 0.2.6 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> stLearn
          

          
          </a>

          
            
            
              <div class="version">
                0.2.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Content:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../stSME_clustering.html">stSME clustering tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../stSME_comparison.html">This tutorial compares clustering performance of three scenarios:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Pseudo-space-time-tutorial.html">Spatial trajectory inference analysis tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci.html">Spatial cell-cell interaction tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_within-spot_single_lr.html">Scenario 1: Cell-cell interaction analysis (single L-R pair, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_within-spot_cpdb_lrs.html">Scenario 2: Cell-cell interaction analysis (CPDB LR pairs, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_between-spot_single_lr.html">Scenario 3: Cell-cell interaction analysis (single L-R pair, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_between-spot_cpdb_lrs.html">Scenario 4: Cell-cell interaction analysis (CPDB L-R pairs, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Working-with-Old-Spatial-Transcriptomics-data.html">Working with Old Spatial Transcriptomics data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Read_slideseq.html">Working with slide-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Read_MERFISH.html">Working with MERFISH data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Read_seqfish.html">Working with SeqFish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ST_deconvolution_visualization.html">Spatial transcriptomics deconvolution visualization</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">stLearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>stlearn.preprocessing.log_scale</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for stlearn.preprocessing.log_scale</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Collection</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Iterable</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">AnnData</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span><span class="p">,</span> <span class="n">isspmatrix_csr</span><span class="p">,</span> <span class="n">csr_matrix</span><span class="p">,</span> <span class="n">spmatrix</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span>
<span class="kn">from</span> <span class="nn">stlearn</span> <span class="kn">import</span> <span class="n">logging</span> <span class="k">as</span> <span class="n">logg</span>
<span class="kn">import</span> <span class="nn">scanpy</span>

<div class="viewcode-block" id="log1p"><a class="viewcode-back" href="../../../stlearn.pp.log1p.html#stlearn.pp.log1p">[docs]</a><span class="k">def</span> <span class="nf">log1p</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">spmatrix</span><span class="p">],</span>
    <span class="n">copy</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">chunked</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">chunk_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">AnnData</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;\</span>
<span class="sd">    Wrap function of scanpy.pp.log1p</span>
<span class="sd">    Copyright (c) 2017 F. Alexander Wolf, P. Angerer, Theis Lab</span>

<span class="sd">    Logarithmize the data matrix.</span>
<span class="sd">    Computes :math:`X = \\log(X + 1)`,</span>
<span class="sd">    where :math:`log` denotes the natural logarithm unless a different base is given.</span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data</span>
<span class="sd">        The (annotated) data matrix of shape `n_obs` × `n_vars`.</span>
<span class="sd">        Rows correspond to cells and columns to genes.</span>
<span class="sd">    copy</span>
<span class="sd">        If an :class:`~anndata.AnnData` is passed, determines whether a copy</span>
<span class="sd">        is returned.</span>
<span class="sd">    chunked</span>
<span class="sd">        Process the data matrix in chunks, which will save memory.</span>
<span class="sd">        Applies only to :class:`~anndata.AnnData`.</span>
<span class="sd">    chunk_size</span>
<span class="sd">        `n_obs` of the chunks to process the data in.</span>
<span class="sd">    base</span>
<span class="sd">        Base of the logarithm. Natural logarithm is used by default.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Returns or updates `data`, depending on `copy`.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="n">copy</span><span class="p">,</span> <span class="n">chunked</span><span class="o">=</span><span class="n">chunked</span><span class="p">,</span> <span class="n">chunk_size</span><span class="o">=</span><span class="n">chunk_size</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="n">base</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Log transformation step is finished in adata.X&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="scale"><a class="viewcode-back" href="../../../stlearn.pp.scale.html#stlearn.pp.scale">[docs]</a><span class="k">def</span> <span class="nf">scale</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">spmatrix</span><span class="p">],</span>
    <span class="n">zero_center</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">max_value</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">copy</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">AnnData</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;\</span>
<span class="sd">    Wrap function of scanpy.pp.scale</span>

<span class="sd">    Scale data to unit variance and zero mean.</span>
<span class="sd">    .. note::</span>
<span class="sd">        Variables (genes) that do not display any variation (are constant across</span>
<span class="sd">        all observations) are retained and set to 0 during this operation. In</span>
<span class="sd">        the future, they might be set to NaNs.</span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data</span>
<span class="sd">        The (annotated) data matrix of shape `n_obs` × `n_vars`.</span>
<span class="sd">        Rows correspond to cells and columns to genes.</span>
<span class="sd">    zero_center</span>
<span class="sd">        If `False`, omit zero-centering variables, which allows to handle sparse</span>
<span class="sd">        input efficiently.</span>
<span class="sd">    max_value</span>
<span class="sd">        Clip (truncate) to this value after scaling. If `None`, do not clip.</span>
<span class="sd">    copy</span>
<span class="sd">        If an :class:`~anndata.AnnData` is passed,</span>
<span class="sd">        determines whether a copy is returned.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Depending on `copy` returns or updates `adata` with a scaled `adata.X`.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">zero_center</span><span class="o">=</span><span class="n">zero_center</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="n">max_value</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="n">copy</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Scale step is finished in adata.X&quot;</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Genomics and Machine Learning lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>