

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>stlearn.preprocessing.normalize &mdash; stLearn 0.2.7 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/custom.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> stLearn
          

          
          </a>

          
            
            
              <div class="version">
                0.2.7
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Content:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../stlearn-interactive-tutorial.html">stLearn interactive web app tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../stSME_clustering.html">stSME clustering tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../stSME_comparison.html">StSME comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Pseudo-space-time-tutorial.html">Spatial trajectory inference analysis tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci.html">Spatial cell-cell interaction tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_within-spot_single_lr.html">Scenario 1: Cell-cell interaction analysis (single L-R pair, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_within-spot_cpdb_lrs.html">Scenario 2: Cell-cell interaction analysis (CPDB LR pairs, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_between-spot_single_lr.html">Scenario 3: Cell-cell interaction analysis (single L-R pair, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cci_between-spot_cpdb_lrs.html">Scenario 4: Cell-cell interaction analysis (CPDB L-R pairs, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Working-with-Old-Spatial-Transcriptomics-data.html">Working with Old Spatial Transcriptomics data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Read_slideseq.html">Working with slide-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Read_MERFISH.html">Working with MERFISH data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Read_seqfish.html">Working with SeqFish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ST_deconvolution_visualization.html">Spatial transcriptomics deconvolution visualization</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">stLearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>stlearn.preprocessing.normalize</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for stlearn.preprocessing.normalize</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">Dict</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">AnnData</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span>
<span class="kn">from</span> <span class="nn">sklearn.utils</span> <span class="kn">import</span> <span class="n">sparsefuncs</span>
<span class="kn">from</span> <span class="nn">stlearn._compat</span> <span class="kn">import</span> <span class="n">Literal</span>
<span class="kn">import</span> <span class="nn">scanpy</span>

<div class="viewcode-block" id="normalize_total"><a class="viewcode-back" href="../../../stlearn.pp.normalize_total.html#stlearn.pp.normalize_total">[docs]</a><span class="k">def</span> <span class="nf">normalize_total</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">target_sum</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">exclude_highly_expressed</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">max_fraction</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">key_added</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">layers</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;all&#39;</span><span class="p">],</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">layer_norm</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">inplace</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;\</span>
<span class="sd">    Wrap function from scanpy.pp.log1p</span>
<span class="sd">    Normalize counts per cell.</span>
<span class="sd">    If choosing `target_sum=1e6`, this is CPM normalization.</span>
<span class="sd">    If `exclude_highly_expressed=True`, very highly expressed genes are excluded</span>
<span class="sd">    from the computation of the normalization factor (size factor) for each</span>
<span class="sd">    cell. This is meaningful as these can strongly influence the resulting</span>
<span class="sd">    normalized values for all other genes [Weinreb17]_.</span>
<span class="sd">    Similar functions are used, for example, by Seurat [Satija15]_, Cell Ranger</span>
<span class="sd">    [Zheng17]_ or SPRING [Weinreb17]_.</span>
<span class="sd">    Params</span>
<span class="sd">    ------</span>
<span class="sd">    adata</span>
<span class="sd">        The annotated data matrix of shape `n_obs` Ã— `n_vars`.</span>
<span class="sd">        Rows correspond to cells and columns to genes.</span>
<span class="sd">    target_sum</span>
<span class="sd">        If `None`, after normalization, each observation (cell) has a total</span>
<span class="sd">        count equal to the median of total counts for observations (cells)</span>
<span class="sd">        before normalization.</span>
<span class="sd">    exclude_highly_expressed</span>
<span class="sd">        Exclude (very) highly expressed genes for the computation of the</span>
<span class="sd">        normalization factor (size factor) for each cell. A gene is considered</span>
<span class="sd">        highly expressed, if it has more than `max_fraction` of the total counts</span>
<span class="sd">        in at least one cell. The not-excluded genes will sum up to</span>
<span class="sd">        `target_sum`.</span>
<span class="sd">    max_fraction</span>
<span class="sd">        If `exclude_highly_expressed=True`, consider cells as highly expressed</span>
<span class="sd">        that have more counts than `max_fraction` of the original total counts</span>
<span class="sd">        in at least one cell.</span>
<span class="sd">    key_added</span>
<span class="sd">        Name of the field in `adata.obs` where the normalization factor is</span>
<span class="sd">        stored.</span>
<span class="sd">    layers</span>
<span class="sd">        List of layers to normalize. Set to `&#39;all&#39;` to normalize all layers.</span>
<span class="sd">    layer_norm</span>
<span class="sd">        Specifies how to normalize layers:</span>
<span class="sd">        * If `None`, after normalization, for each layer in *layers* each cell</span>
<span class="sd">          has a total count equal to the median of the *counts_per_cell* before</span>
<span class="sd">          normalization of the layer.</span>
<span class="sd">        * If `&#39;after&#39;`, for each layer in *layers* each cell has</span>
<span class="sd">          a total count equal to `target_sum`.</span>
<span class="sd">        * If `&#39;X&#39;`, for each layer in *layers* each cell has a total count</span>
<span class="sd">          equal to the median of total counts for observations (cells) of</span>
<span class="sd">          `adata.X` before normalization.</span>
<span class="sd">    inplace</span>
<span class="sd">        Whether to update `adata` or return dictionary with normalized copies of</span>
<span class="sd">        `adata.X` and `adata.layers`.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Returns dictionary with normalized copies of `adata.X` and `adata.layers`</span>
<span class="sd">    or updates `adata` with normalized version of the original</span>
<span class="sd">    `adata.X` and `adata.layers`, depending on `inplace`.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="n">target_sum</span><span class="p">,</span>
                    <span class="n">exclude_highly_expressed</span><span class="o">=</span><span class="n">exclude_highly_expressed</span><span class="p">,</span>
                    <span class="n">max_fraction</span><span class="o">=</span><span class="n">max_fraction</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="n">key_added</span><span class="p">,</span>
                    <span class="n">layers</span><span class="o">=</span><span class="n">layers</span><span class="p">,</span> <span class="n">layer_norm</span><span class="o">=</span><span class="n">layer_norm</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="n">inplace</span><span class="p">)</span>

    <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;normalized_total&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Normalization step is finished in adata.X&quot;</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Genomics and Machine Learning lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>