

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Spatial trajectory inference analysis tutorial &mdash; stLearn 0.3.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Cell-cell interaction analysis (single LR, between-spot)" href="cci_single-lr_between-spot.html" />
    <link rel="prev" title="stSME comparison" href="stSME_comparison.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> stLearn
          

          
          </a>

          
            
            
              <div class="version">
                0.3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Content:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="stlearn-interactive-tutorial.html">stLearn interactive web app tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="stSME_clustering.html">stSME clustering tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="stSME_comparison.html">stSME comparison</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Spatial trajectory inference analysis tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#1.-Preparation">1. Preparation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Read-and-preprocess-data">Read and preprocess data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Clustering-data">Clustering data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#2.-Spatial-trajectory-inference">2. Spatial trajectory inference</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Choosing-root">Choosing root</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Spatial-trajectory-inference---global-level">Spatial trajectory inference - global level</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Transition-markers-detection">Transition markers detection</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cci_single-lr_between-spot.html">Cell-cell interaction analysis (single LR, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_single-lr_within-spot.html">Cell-cell interaction analysis (single LR, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_multiple-lr_between-spot.html">Cell-cell interaction analysis (multiple LRs, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_multiple-lr_within-spot.html">Cell-cell interaction analysis (multiple LRs, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Working-with-Old-Spatial-Transcriptomics-data.html">Working with Old Spatial Transcriptomics data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_slideseq.html">Working with slide-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_MERFISH.html">Working with MERFISH data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_seqfish.html">Working with SeqFish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="ST_deconvolution_visualization.html">Spatial transcriptomics deconvolution visualization</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stLearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Spatial trajectory inference analysis tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/Pseudo-time-space-tutorial.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    min-width: 5ex;
    padding-top: 0.3rem;
    padding-right: 0.3rem;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 0.3rem;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Spatial-trajectory-inference-analysis-tutorial">
<h1>Spatial trajectory inference analysis tutorial<a class="headerlink" href="#Spatial-trajectory-inference-analysis-tutorial" title="Permalink to this headline">¶</a></h1>
<p>In this tutorial, we are using both spatial information and gene expression profile to perform spatial trajectory inference to explore the progression of Ductal Carcinoma in situ (DCIS) - Invasive Ductal Cancer (IDC)</p>
<p>Source: <a class="reference external" href="https://support.10xgenomics.com/spatial-gene-expression/datasets">https://support.10xgenomics.com/spatial-gene-expression/datasets</a></p>
<div class="section" id="1.-Preparation">
<h2>1. Preparation<a class="headerlink" href="#1.-Preparation" title="Permalink to this headline">¶</a></h2>
<p>We are trying to keep it focus on spatial trajectory inference then every step from reading data, processing to clustering, we will give the code here to easier for user to use.</p>
<div class="section" id="Read-and-preprocess-data">
<h3>Read and preprocess data<a class="headerlink" href="#Read-and-preprocess-data" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">stlearn</span> <span class="k">as</span> <span class="nn">st</span>
<span class="n">st</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">120</span><span class="p">)</span>
<span class="c1"># Reading data</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">Read10X</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;BCBA_10x_data_Path&quot;</span><span class="p">)</span>
<span class="c1"># Save raw_count</span>
<span class="n">data</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;raw_count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">X</span>
<span class="c1"># Preprocessing</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="c1"># Keep raw data</span>
<span class="n">data</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">data</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Normalization step is finished in adata.X
Log transformation step is finished in adata.X
Scale step is finished in adata.X
</pre></div></div>
</div>
</div>
<div class="section" id="Clustering-data">
<h3>Clustering data<a class="headerlink" href="#Clustering-data" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Run PCA</span>
<span class="n">st</span><span class="o">.</span><span class="n">em</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">n_comps</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="c1"># Tiling image</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">tiling</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">out_path</span><span class="o">=</span><span class="s2">&quot;tiling&quot;</span><span class="p">,</span><span class="n">crop_size</span> <span class="o">=</span> <span class="mi">40</span><span class="p">)</span>
<span class="c1"># Using Deep Learning to extract feature</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="c1"># Apply stSME spatial-PCA option</span>
<span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">morphology</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span><span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span><span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca_morphology&#39;</span><span class="p">,</span><span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">louvain</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PCA is done! Generated in adata.obsm[&#39;X_pca&#39;], adata.uns[&#39;pca&#39;] and adata.varm[&#39;PCs&#39;]
The morphology feature is added to adata.obsm[&#39;X_morphology&#39;]!
The data adjusted by morphology is added to adata.obsm[&#39;X_pca_morphology&#39;]
Created k-Nearest-Neighbor graph in adata.uns[&#39;neighbors&#39;]
Applying Louvain clustering ...
Louvain clustering is done! The labels are stored in adata.obs[&#39;louvain&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span><span class="n">tissue_alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">spot_size</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span><span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_8_0.png"><img alt="_images/Pseudo-time-space-tutorial_8_0.png" src="_images/Pseudo-time-space-tutorial_8_0.png" style="width: 443px; height: 386px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">add</span><span class="o">.</span><span class="n">annotation</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">label_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Fatty tissue,immune/lymphoid 1 MALAT1+&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Invasive cancer,fibrous tissue 1 CXCL14+&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Invasive cancer,fibrous tissue 2 CRISP3+&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Invasive cancer,fibrous tissue, fatty tissue&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Fatty tissue,immune/lymphoid 2 IGKC+&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Fibrous tissue&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Invasive cancer,fibrous tissue (DCIS)&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Fatty tissue, Fibrous tissue&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Invasive cancer,immune/lymphoid (IDC)&#39;</span> <span class="p">,</span>
                                   <span class="s1">&#39;Invasive cancer,fatty tissue 3 MUC5B+&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;Fatty tissue&#39;</span><span class="p">],</span>
                 <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;louvain_anno&quot;</span><span class="p">,</span><span class="n">tissue_alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">spot_size</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span><span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The annotation is added to adata.obs[&#39;louvain_anno&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_9_1.png"><img alt="_images/Pseudo-time-space-tutorial_9_1.png" src="_images/Pseudo-time-space-tutorial_9_1.png" style="width: 788px; height: 386px;" /></a>
</div>
</div>
</div>
</div>
<div class="section" id="2.-Spatial-trajectory-inference">
<h2>2. Spatial trajectory inference<a class="headerlink" href="#2.-Spatial-trajectory-inference" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Choosing-root">
<h3>Choosing root<a class="headerlink" href="#Choosing-root" title="Permalink to this headline">¶</a></h3>
<p>3733 is the index of the spot that we chose as root. It in the DCIS cluster (6). We recommend the root spot should be in the end/begin of a cluster in UMAP space. You can find min/max point of a cluster in UMAP as root.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">data</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;iroot&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">3733</span>
<span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">pseudotime</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">eps</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span><span class="n">use_sme</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Created k-Nearest-Neighbor graph in adata.uns[&#39;neighbors&#39;]
The data adjusted by morphology is added to adata.obsm[&#39;X_diffmap_morphology&#39;]
</pre></div></div>
</div>
</div>
<div class="section" id="Spatial-trajectory-inference---global-level">
<h3>Spatial trajectory inference - global level<a class="headerlink" href="#Spatial-trajectory-inference---global-level" title="Permalink to this headline">¶</a></h3>
<p>We run the global level of pseudo-time-space (PSTS) method to reconstruct the spatial trajectory between cluster 6 (DCIS) and 8 (lesions IDC)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">pseudotimespace_global</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span><span class="n">list_cluster</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">])</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span><span class="n">show_trajectory</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">list_cluster</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">],</span><span class="n">show_subcluster</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Screening PTS global graph...
Calculate the graph dissimilarity using Laplacian matrix...
The optimized weighting is: 0.35
Start to construct the trajectory: 6 -&gt; 8
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_14_1.png"><img alt="_images/Pseudo-time-space-tutorial_14_1.png" src="_images/Pseudo-time-space-tutorial_14_1.png" style="width: 443px; height: 386px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">tree_plot</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_15_0.png"><img alt="_images/Pseudo-time-space-tutorial_15_0.png" src="_images/Pseudo-time-space-tutorial_15_0.png" style="width: 953px; height: 396px;" /></a>
</div>
</div>
</div>
<div class="section" id="Transition-markers-detection">
<h3>Transition markers detection<a class="headerlink" href="#Transition-markers-detection" title="Permalink to this headline">¶</a></h3>
<p>Based on the spatial trajectory/tree plot, we can see 2 clades are started from sub-cluster 6 and 13. Then we run the function to detect the highly correlated genes with the PSTS values.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">detect_transition_markers_clades</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">clade</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span><span class="n">use_raw_count</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">cutoff_spearman</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Detecting the transition markers of clade_6...
Transition markers result is stored in adata.uns[&#39;clade_6&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">detect_transition_markers_clades</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">clade</span><span class="o">=</span><span class="mi">13</span><span class="p">,</span><span class="n">use_raw_count</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">cutoff_spearman</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Detecting the transition markers of clade_13...
Transition markers result is stored in adata.uns[&#39;clade_13&#39;]
</pre></div></div>
</div>
<p>For the transition markers plot, genes from left side (red) are negatively correlated with the spatial trajectory and genes from right side (blue) are positively correlated with the spatial trajectory.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">transition_markers_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">top_genes</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="n">trajectory</span><span class="o">=</span><span class="s2">&quot;clade_6&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_20_0.png"><img alt="_images/Pseudo-time-space-tutorial_20_0.png" src="_images/Pseudo-time-space-tutorial_20_0.png" style="width: 773px; height: 572px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">transition_markers_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">top_genes</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="n">trajectory</span><span class="o">=</span><span class="s2">&quot;clade_13&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_21_0.png"><img alt="_images/Pseudo-time-space-tutorial_21_0.png" src="_images/Pseudo-time-space-tutorial_21_0.png" style="width: 785px; height: 572px;" /></a>
</div>
</div>
<p>We also provide a function to compare the transition markers between two clades.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">compare_transitions</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">trajectories</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;clade_6&quot;</span><span class="p">,</span><span class="s2">&quot;clade_13&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The result of comparison between clade_6 and clade_13 stored in &#39;adata.uns[&#39;compare_result&#39;]&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">trajectory</span><span class="o">.</span><span class="n">DE_transition_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">top_genes</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_24_0.png"><img alt="_images/Pseudo-time-space-tutorial_24_0.png" src="_images/Pseudo-time-space-tutorial_24_0.png" style="width: 818px; height: 568px;" /></a>
</div>
</div>
<p>We can visualize some genes that different between two clades.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">genes</span><span class="o">=</span><span class="s2">&quot;CTPS2&quot;</span><span class="p">,</span><span class="n">list_clusters</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_26_0.png"><img alt="_images/Pseudo-time-space-tutorial_26_0.png" src="_images/Pseudo-time-space-tutorial_26_0.png" style="width: 573px; height: 483px;" /></a>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">genes</span><span class="o">=</span><span class="s2">&quot;IFITM3&quot;</span><span class="p">,</span><span class="n">list_clusters</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<a class="reference internal image-reference" href="_images/Pseudo-time-space-tutorial_27_0.png"><img alt="_images/Pseudo-time-space-tutorial_27_0.png" src="_images/Pseudo-time-space-tutorial_27_0.png" style="width: 573px; height: 483px;" /></a>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="cci_single-lr_between-spot.html" class="btn btn-neutral float-right" title="Cell-cell interaction analysis (single LR, between-spot)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="stSME_comparison.html" class="btn btn-neutral float-left" title="stSME comparison" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Genomics and Machine Learning lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>