

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>stlearn.pp.filter_genes &mdash; stLearn 0.2.6 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="stlearn.pp.log1p" href="stlearn.pp.log1p.html" />
    <link rel="prev" title="stlearn.add.annotation" href="stlearn.add.annotation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> stLearn
          

          
          </a>

          
            
            
              <div class="version">
                0.2.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Content:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="api.html#module-stlearn.wrapper">Wrapper functions: <cite>wrapper</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-stlearn.add">Add: <cite>add</cite></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="api.html#module-stlearn.pp">Preprocessing: <cite>pp</cite></a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">stlearn.pp.filter_genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="stlearn.pp.log1p.html">stlearn.pp.log1p</a></li>
<li class="toctree-l3"><a class="reference internal" href="stlearn.pp.normalize_total.html">stlearn.pp.normalize_total</a></li>
<li class="toctree-l3"><a class="reference internal" href="stlearn.pp.scale.html">stlearn.pp.scale</a></li>
<li class="toctree-l3"><a class="reference internal" href="stlearn.pp.neighbors.html">stlearn.pp.neighbors</a></li>
<li class="toctree-l3"><a class="reference internal" href="stlearn.pp.tiling.html">stlearn.pp.tiling</a></li>
<li class="toctree-l3"><a class="reference internal" href="stlearn.pp.extract_feature.html">stlearn.pp.extract_feature</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-stlearn.em">Embedding: <cite>em</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-stlearn.spatial">Spatial: <cite>spatial</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-stlearn.tl">Tools: <cite>tl</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-stlearn.pl">Plot: <cite>pl</cite></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="stSME_clustering.html">stSME clustering tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="stSME_comparison.html">This tutorial compares clustering performance of three scenarios:</a></li>
<li class="toctree-l1"><a class="reference internal" href="Pseudo-space-time-tutorial.html">Spatial trajectory inference analysis tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci.html">Spatial cell-cell interaction tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_within-spot_single_lr.html">Scenario 1: Cell-cell interaction analysis (single L-R pair, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_within-spot_cpdb_lrs.html">Scenario 2: Cell-cell interaction analysis (CPDB LR pairs, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_between-spot_single_lr.html">Scenario 3: Cell-cell interaction analysis (single L-R pair, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_between-spot_cpdb_lrs.html">Scenario 4: Cell-cell interaction analysis (CPDB L-R pairs, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Working-with-Old-Spatial-Transcriptomics-data.html">Working with Old Spatial Transcriptomics data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_slideseq.html">Working with slide-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_MERFISH.html">Working with MERFISH data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_seqfish.html">Working with SeqFish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="ST_deconvolution_visualization.html">Spatial transcriptomics deconvolution visualization</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stLearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="api.html">API</a> &raquo;</li>
        
      <li>stlearn.pp.filter_genes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/stlearn.pp.filter_genes.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="stlearn-pp-filter-genes">
<h1>stlearn.pp.filter_genes<a class="headerlink" href="#stlearn-pp-filter-genes" title="Permalink to this headline">Â¶</a></h1>
<dl class="py function">
<dt id="stlearn.pp.filter_genes">
<code class="sig-prename descclassname">stlearn.pp.</code><code class="sig-name descname">filter_genes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span><span class="p">:</span> <span class="n">anndata._core.anndata.AnnData</span></em>, <em class="sig-param"><span class="n">min_counts</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_counts</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_cells</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span> &#x2192; Union<span class="p">[</span>anndata._core.anndata.AnnData<span class="p">, </span>None<span class="p">, </span>Tuple<span class="p">[</span>numpy.ndarray<span class="p">, </span>numpy.ndarray<span class="p">]</span><span class="p">]</span><a class="reference internal" href="_modules/stlearn/preprocessing/filter_genes.html#filter_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#stlearn.pp.filter_genes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Wrap function scanpy.pp.filter_genes</p>
<p>Filter genes based on number of cells or counts.
Keep genes that have at least <cite>min_counts</cite> counts or are expressed in at
least <cite>min_cells</cite> cells or have at most <cite>max_counts</cite> counts or are expressed
in at most <cite>max_cells</cite> cells.
Only provide one of the optional parameters <cite>min_counts</cite>, <cite>min_cells</cite>,
<cite>max_counts</cite>, <cite>max_cells</cite> per call.
:param data: An annotated data matrix of shape <cite>n_obs</cite> Ã <cite>n_vars</cite>. Rows correspond</p>
<blockquote>
<div><p>to cells and columns to genes.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_counts</strong> â Minimum number of counts required for a gene to pass filtering.</p></li>
<li><p><strong>min_cells</strong> â Minimum number of cells expressed required for a gene to pass filtering.</p></li>
<li><p><strong>max_counts</strong> â Maximum number of counts required for a gene to pass filtering.</p></li>
<li><p><strong>max_cells</strong> â Maximum number of cells expressed required for a gene to pass filtering.</p></li>
<li><p><strong>inplace</strong> â Perform computation inplace or return result.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p>Depending on <cite>inplace</cite>, returns the following arrays or directly subsets</p></li>
<li><p><em>and annotates the data matrix</em></p></li>
<li><p><em>gene_subset</em> â Boolean index mask that does filtering. <cite>True</cite> means that the
gene is kept. <cite>False</cite> means the gene is removed.</p></li>
<li><p><em>number_per_gene</em> â Depending on what was tresholded (<cite>counts</cite> or <cite>cells</cite>), the array stores
<cite>n_counts</cite> or <cite>n_cells</cite> per gene.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="stlearn.pp.log1p.html" class="btn btn-neutral float-right" title="stlearn.pp.log1p" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="stlearn.add.annotation.html" class="btn btn-neutral float-left" title="stlearn.add.annotation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Genomics and Machine Learning lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>