

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>This tutorial compares clustering performance of three scenarios: &mdash; stLearn 0.2.6 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Spatial trajectory inference analysis tutorial" href="Pseudo-space-time-tutorial.html" />
    <link rel="prev" title="stSME clustering tutorial" href="stSME_clustering.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> stLearn
          

          
          </a>

          
            
            
              <div class="version">
                0.2.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Content:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="stSME_clustering.html">stSME clustering tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">This tutorial compares clustering performance of three scenarios:</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#We-ues-Human-Brain-dorsolateral-prefrontal-cortex-(DLPFC)-Visium-dataset-from-here">We ues Human Brain dorsolateral prefrontal cortex (DLPFC) Visium dataset from here</a></li>
<li class="toctree-l2"><a class="reference internal" href="#(1)-normal-(without-stSME)">(1) normal (without stSME)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#(2)-stSME-applied-on-dimension-reduced-spaces-(PCA)-as-normalization">(2) stSME applied on dimension reduced spaces (PCA) as normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#(3)-stSME-applied-on-raw-gene-counts-as-imputation">(3) stSME applied on raw gene counts as imputation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Pseudo-space-time-tutorial.html">Spatial trajectory inference analysis tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci.html">Spatial cell-cell interaction tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_within-spot_single_lr.html">Scenario 1: Cell-cell interaction analysis (single L-R pair, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_within-spot_cpdb_lrs.html">Scenario 2: Cell-cell interaction analysis (CPDB LR pairs, within-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_between-spot_single_lr.html">Scenario 3: Cell-cell interaction analysis (single L-R pair, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="cci_between-spot_cpdb_lrs.html">Scenario 4: Cell-cell interaction analysis (CPDB L-R pairs, between-spot)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Working-with-Old-Spatial-Transcriptomics-data.html">Working with Old Spatial Transcriptomics data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_slideseq.html">Working with slide-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_MERFISH.html">Working with MERFISH data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Read_seqfish.html">Working with SeqFish data</a></li>
<li class="toctree-l1"><a class="reference internal" href="ST_deconvolution_visualization.html">Spatial transcriptomics deconvolution visualization</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">stLearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>This tutorial compares clustering performance of three scenarios:</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/stSME_comparison.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="This-tutorial-compares-clustering-performance-of-three-scenarios:">
<h1>This tutorial compares clustering performance of three scenarios:<a class="headerlink" href="#This-tutorial-compares-clustering-performance-of-three-scenarios:" title="Permalink to this headline">¶</a></h1>
<ol class="arabic simple">
<li><p>normal (without stSME)</p></li>
<li><p>stSME applied on dimension reduced spaces (PCA) as normalization and</p></li>
<li><p>stSME applied on raw gene counts as imputation</p></li>
</ol>
<div class="section" id="We-ues-Human-Brain-dorsolateral-prefrontal-cortex-(DLPFC)-Visium-dataset-from-here">
<h2>We ues Human Brain dorsolateral prefrontal cortex (DLPFC) Visium dataset from <a class="reference external" href="https://github.com/LieberInstitute/HumanPilot">here</a><a class="headerlink" href="#We-ues-Human-Brain-dorsolateral-prefrontal-cortex-(DLPFC)-Visium-dataset-from-here" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load modules</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">adjusted_rand_score</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">LabelEncoder</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scanpy</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">stlearn</span> <span class="k">as</span> <span class="nn">st</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># specify parameters</span>
<span class="n">tile_size</span> <span class="o">=</span> <span class="mi">40</span>
<span class="n">radius</span> <span class="o">=</span> <span class="mf">90.0</span>
<span class="n">rates</span> <span class="o">=</span> <span class="mi">5</span>

<span class="c1"># all samples in this dataset</span>
<span class="n">sample_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;151507&quot;</span><span class="p">,</span> <span class="s2">&quot;151508&quot;</span><span class="p">,</span> <span class="s2">&quot;151509&quot;</span><span class="p">,</span>
               <span class="s2">&quot;151510&quot;</span><span class="p">,</span> <span class="s2">&quot;151669&quot;</span><span class="p">,</span> <span class="s2">&quot;151670&quot;</span><span class="p">,</span>
               <span class="s2">&quot;151671&quot;</span><span class="p">,</span> <span class="s2">&quot;151672&quot;</span><span class="p">,</span> <span class="s2">&quot;151673&quot;</span><span class="p">,</span>
               <span class="s2">&quot;151674&quot;</span><span class="p">,</span> <span class="s2">&quot;151675&quot;</span><span class="p">,</span> <span class="s2">&quot;151676&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># specify PATH</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">sample</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="n">BASE_PATH</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/90days/uqxtan9/Human_Brain_spatialLIBD&quot;</span><span class="p">)</span>
<span class="n">MARKER_GENE_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="s2">&quot;marker_gene.tsv&quot;</span>
<span class="n">SAMPLE_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="n">sample</span>
<span class="n">CM_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="n">sample</span> <span class="o">/</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_filtered_feature_bc_matrix.h5&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
<span class="n">IMG_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="n">sample</span> <span class="o">/</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_tissue_hires_image.png&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
<span class="n">SPOT_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="n">sample</span> <span class="o">/</span> <span class="s2">&quot;tissue_positions_list.txt&quot;</span>
<span class="n">SCALE_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="n">sample</span> <span class="o">/</span> <span class="s2">&quot;scalefactors_json.json&quot;</span>
<span class="n">TILE_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_tiles&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
<span class="n">TILE_PATH</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">MARKER_GENE_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="s2">&quot;marker_gene.tsv&quot;</span>
<span class="n">GROUND_TRUTH_PATH</span> <span class="o">=</span> <span class="n">BASE_PATH</span> <span class="o">/</span> <span class="n">sample</span> <span class="o">/</span> <span class="s2">&quot;cluster_labels_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># load ST data and ground truth label for each spot</span>
<span class="n">ground_truth_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">GROUND_TRUTH_PATH</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">ground_truth_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">ground_truth_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">7</span><span class="p">:])</span>

<span class="n">marker_gene_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">MARKER_GENE_PATH</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">marker_gene</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">marker_gene_df</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()]</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">Read10X</span><span class="p">(</span><span class="n">SAMPLE_PATH</span><span class="p">)</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">marker_gene</span><span class="p">]</span>

<span class="n">ground_truth_df</span> <span class="o">=</span> <span class="n">ground_truth_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>

<span class="n">n_cluster</span> <span class="o">=</span> <span class="nb">len</span><span class="p">((</span><span class="nb">set</span><span class="p">(</span><span class="n">ground_truth_df</span><span class="p">[</span><span class="s2">&quot;ground_truth&quot;</span><span class="p">])))</span>

<span class="n">data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;ground_truth&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ground_truth_df</span><span class="p">[</span><span class="s2">&quot;ground_truth&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># ground truth label</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;ground_truth&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_6_0.png" src="_images/stSME_comparison_6_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># pre-processing for gene count table</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Some cells have total count of genes equal to zero
Normalization step is finished in adata.X
Log transformation step is finished in adata.X
Scale step is finished in adata.X
</pre></div></div>
</div>
</div>
<div class="section" id="(1)-normal-(without-stSME)">
<h2>(1) normal (without stSME)<a class="headerlink" href="#(1)-normal-(without-stSME)" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># run PCA for gene expression data</span>
<span class="n">st</span><span class="o">.</span><span class="n">em</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PCA is done! Generated in adata.obsm[&#39;X_pca&#39;], adata.uns[&#39;pca&#39;] and adata.varm[&#39;PCs&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># K-means clustering on PCA</span>
<span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">kmeans</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_clusters</span><span class="o">=</span><span class="n">n_cluster</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_pca_kmeans&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;X_pca_kmeans&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Applying Kmeans clustering ...
Kmeans clustering is done! The labels are stored in adata.obs[&#34;kmeans&#34;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_10_1.png" src="_images/stSME_comparison_10_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># K-means clustering on UMAP</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">em</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">kmeans</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_clusters</span><span class="o">=</span><span class="n">n_cluster</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_umap&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_umap_kmeans&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;X_umap_kmeans&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Created k-Nearest-Neighbor graph in adata.uns[&#39;neighbors&#39;]
UMAP is done! Generated in adata.obsm[&#39;X_umap&#39;] nad adata.uns[&#39;umap&#39;]
Applying Kmeans clustering ...
Kmeans clustering is done! The labels are stored in adata.obs[&#34;kmeans&#34;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_11_1.png" src="_images/stSME_comparison_11_1.png" />
</div>
</div>
</div>
<div class="section" id="(2)-stSME-applied-on-dimension-reduced-spaces-(PCA)-as-normalization">
<h2>(2) stSME applied on dimension reduced spaces (PCA) as normalization<a class="headerlink" href="#(2)-stSME-applied-on-dimension-reduced-spaces-(PCA)-as-normalization" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># pre-processing for spot image</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">tiling</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">TILE_PATH</span><span class="p">,</span> <span class="n">crop_size</span><span class="o">=</span><span class="n">tile_size</span><span class="p">)</span>

<span class="c1"># this step uses deep learning model to extract high-level features from tile images</span>
<span class="c1"># may need few minutes to be completed</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="c1"># apply stSME to normalise PCA</span>
<span class="c1"># radius=90 includes three units of surrounding spots</span>
<span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">morphology</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">radius</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
                             <span class="n">similarity_matrix</span><span class="o">=</span><span class="s2">&quot;cosine&quot;</span><span class="p">,</span> <span class="n">rates</span><span class="o">=</span><span class="n">rates</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca_morphology&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The morphology feature is added to adata.obsm[&#39;X_morphology&#39;]!
The data adjusted by morphology is added to adata.obsm[&#39;X_pca_morphology&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># K-means clustering on stSME normalised PCA</span>
<span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">kmeans</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_clusters</span><span class="o">=</span><span class="n">n_cluster</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_pca_kmeans&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;X_pca_kmeans&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Applying Kmeans clustering ...
Kmeans clustering is done! The labels are stored in adata.obs[&#34;kmeans&#34;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_14_1.png" src="_images/stSME_comparison_14_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># K-means clustering on stSME normalised UMAP</span>
<span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">em</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">kmeans</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_clusters</span><span class="o">=</span><span class="n">n_cluster</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_umap&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_umap_kmeans&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;X_umap_kmeans&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Created k-Nearest-Neighbor graph in adata.uns[&#39;neighbors&#39;]
UMAP is done! Generated in adata.obsm[&#39;X_umap&#39;] nad adata.uns[&#39;umap&#39;]
Applying Kmeans clustering ...
Kmeans clustering is done! The labels are stored in adata.obs[&#34;kmeans&#34;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_15_1.png" src="_images/stSME_comparison_15_1.png" />
</div>
</div>
</div>
<div class="section" id="(3)-stSME-applied-on-raw-gene-counts-as-imputation">
<h2>(3) stSME applied on raw gene counts as imputation<a class="headerlink" href="#(3)-stSME-applied-on-raw-gene-counts-as-imputation" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># apply stSME to raw counts</span>
<span class="c1"># radius=90 includes three units of surrounding spots</span>
<span class="n">st</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">morphology</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;raw&quot;</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">radius</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
                             <span class="n">similarity_matrix</span><span class="o">=</span><span class="s2">&quot;cosine&quot;</span><span class="p">,</span> <span class="n">rates</span><span class="o">=</span><span class="n">rates</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;raw_morphology&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The data adjusted by morphology is added to adata.obsm[&#39;raw_morphology&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">em</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PCA is done! Generated in adata.obsm[&#39;X_pca&#39;], adata.uns[&#39;pca&#39;] and adata.varm[&#39;PCs&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">kmeans</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_clusters</span><span class="o">=</span><span class="n">n_cluster</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_pca_kmeans&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;X_pca_kmeans&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Applying Kmeans clustering ...
Kmeans clustering is done! The labels are stored in adata.obs[&#34;kmeans&#34;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_19_1.png" src="_images/stSME_comparison_19_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">em</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clustering</span><span class="o">.</span><span class="n">kmeans</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_clusters</span><span class="o">=</span><span class="n">n_cluster</span><span class="p">,</span> <span class="n">use_data</span><span class="o">=</span><span class="s2">&quot;X_umap&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_umap_kmeans&quot;</span><span class="p">)</span>
<span class="n">st</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cluster_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">use_label</span><span class="o">=</span><span class="s2">&quot;X_umap_kmeans&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Created k-Nearest-Neighbor graph in adata.uns[&#39;neighbors&#39;]
UMAP is done! Generated in adata.obsm[&#39;X_umap&#39;] nad adata.uns[&#39;umap&#39;]
Applying Kmeans clustering ...
Kmeans clustering is done! The labels are stored in adata.obs[&#34;kmeans&#34;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/stSME_comparison_20_1.png" src="_images/stSME_comparison_20_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Pseudo-space-time-tutorial.html" class="btn btn-neutral float-right" title="Spatial trajectory inference analysis tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="stSME_clustering.html" class="btn btn-neutral float-left" title="stSME clustering tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Genomics and Machine Learning lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>