{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# stLearn interactive web app tutorial\n",
    "\n",
    "## Installation:\n",
    "\n",
    "This is the installation steps for stlearn interactive web app: [Installation](https://github.com/BiomedicalMachineLearning/stlearn_interactive)\n",
    "\n",
    "## Step 1: Read data\n",
    "\n",
    "Set the path in the box as your link to 10X Visium dataset (we will try to expand to other platforms later).\n",
    "\n",
    "![Step 1](_static/img/step1.png)\n",
    "\n",
    "## Step 2: Preprocessing\n",
    "\n",
    "We provided 4 options for the preprocessing step.\n",
    "\n",
    "![Step 2](_static/img/step2.png)\n",
    "\n",
    "## Gene plot\n",
    "\n",
    "After run the preprocessing, the gene plot will be displayed.\n",
    "- Gene: you can input any gene name that exists in the dataset.\n",
    "- Spot alpha: Opacity for the spots.\n",
    "- Tissue alpha: Opacity for the tissue image.\n",
    "- Threshold: the gene expression threshold to be displayed. The spots with larger gene expression than the threshold will be shown.\n",
    "\n",
    "![Gene plot](_static/img/gene_plot.png)\n",
    "\n",
    "## Step 3: Clustering\n",
    "\n",
    "We combine several small steps here. First, we provided:\n",
    "- Number of components to run PCA\n",
    "- You can specify to use stSME or not\n",
    "- Crop size image\n",
    "- Radius to adjust the image for smoothing step\n",
    "- Choosing the method of clustering\n",
    "- Number of the neighbour for KNN\n",
    "\n",
    "![Step 3](_static/img/step3.png)\n",
    "\n",
    "This step will take quite long (if you choose stSME method). You can track the progress in the command line screen.\n",
    "\n",
    "![stSME1](_static/img/extraction.png)\n",
    "\n",
    "![stSME2](_static/img/extraction_done.png)\n",
    "\n",
    "It also generates the initial result of Pseudotime analysis (PAGA method).\n",
    "\n",
    "![PT result](_static/img/PT.png)\n",
    "\n",
    "## Step 4: Pseudotimespace analysis\n",
    "\n",
    "We provided 3 components of the Pseudotimespace analysis in this step.\n",
    "\n",
    "The first one is the mapping of the pseudotimespace value in the Spatial Transcriptomics tissue.\n",
    "\n",
    "Parameters:\n",
    "- Specify the root cluster\n",
    "- Spot index as the root spot\n",
    "- Eps parameter for the localization process\n",
    "- You can choose which clusters should be displayed\n",
    "- Spot alpha and tissue alpha are the opacity options.\n",
    "\n",
    "![PTS result_1](_static/img/PTS_1.png)\n",
    "\n",
    "The second component is the local analysis for PTS.\n",
    "\n",
    "Parameters:\n",
    "- Choose which cluster to run the analysis. This cluster has to be more than one sub-clusters (spatially distributed).\n",
    "- You can choose the weight to balance between gene expression and spatial relationship.\n",
    "- Inverse or reverse the pseudotimespace value.\n",
    "- Spot alpha and tissue alpha are the opacity option.\n",
    "\n",
    "![PTS local](_static/img/PTS_local.png)\n",
    "\n",
    "The final component is the global analysis for PTS.\n",
    "\n",
    "Parameters:\n",
    "- Choose a set of clusters to run the global analysis. With this current algorithm, we recommend choosing a pair of clusters.\n",
    "- You can choose the weight to balance between gene expression and spatial relationship.\n",
    "- Spot alpha and tissue alpha are the opacity option.\n",
    "\n",
    "![PTS global](_static/img/PTS_global.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
